### DISCUSSION TITLE
App does not check storage free space before starting DL

### DISCUSSION THREAD

**Discussion by fcassia2017** [deniscerri/ytdlnis#256] [Q&A] [answered]
HI,

The idea and implementation of this app is very promising. But let me tell you of my usage scenario and a few bugs that make it useless in my case.

Device: Samsung Galaxy Tab A7 Lite (3GB RAM, 32 GB internal storage, 128GB SD card)

OS vsrsion: Android 13,July 2023 patch level

Problem description (short) : 1. App crashes when moved to SD card, 
2. App doesn't check free storage space before starting download
3. App doesn't download directly to SD card. Downloads to internal storage then moves to external folder (if selected) 

Problem description (long) 

Problem: the device is always 90% full in its internal storage, so I have enabled developer mode and proceeded to move ALL APPS to external storage. This allows me to keep between 1GB and 2GB of free internal storage. Running SD Maid Pro, Google Files and the Android settings space cleanup options only go so far, but 2GB free is the best I can get. 

Most apps run fine from external storage and I save all files to the 128GB SD card.

But not with this app. As explained above, the first hurdle:

1. This app crashes after being moved to external storage. I can dig up the error shown and report as a reply to this post
2. The App does not perform a basic FREE SPACE Check before starting a download. So in a system with just 2GB free, it will happily begin downloading a 4.5GB file, only to truncate it in the middle. 
3. This app does Not save directly to the target location. Instead it downloads to internal storage, then copies the file to the target folder in sd external storage. As you can imagine, it fails due to points 2 and 3 and you end up with a truncated download moved to SD card. 

I hope the dev and community can take these points into consideration for future improvements and testing. Before someone tells me to get a better tablet, I can't. And this device is still for sale at Amazon so my guess is that this tablet still has plenty of users out there. This and many other devices with limited internal storage

Cheers, 
FC

**Comment by deniscerri** [comment]
Let me know of the error the pops up when you moved it to sd card.

As per file transfer. Android is pretty restrictive. You can't just download a file directly to sd card. You have to use transfers.

But I'll need to look more into this.
Also there is the issue of knowing what kind of files have downloaded. So the app can create the history item.

If i save in cache i can skim through all files after they are moved and get their paths.

Also caching. Users can pause and cancel downloads. They are stuck in cache and when user resumes it doesn't start from the beginning.

I wonder what can i do

**Comment by deniscerri** [comment]
But on second thought i found a solution so you can download directly to sd with no cache. Will see what i can do

**Reply by fcassia2017** [reply]
HI D,
Thanks for looking into this. I know technically a direct download to SD card must be possible because FDM downloader saves directly to SD card and I was able to download 10GB+ files on this very restricted tablet.

If it helps, I remember reading about the new file access restrictions in Android 11+ and that popular file managers bypass them by targeting Android 9 Api.
https://support.google.com/googleplay/android-developer/answer/10467955?hl=en

I will retry later and post the error message. I might have to record the screen with another phone as everything scrolls very fast. (a debug flag that saves all messages to a file would be nice) 

Thanks again
FC

**Reply by deniscerri** [reply]
I created a build where you can turn off caching in the settings. As you mentioned i used manage external storage flag.

This will work as intended on sd card
But if you have a usb otg it won't work so the app will resort back to caching.

Might release it this weekend or something. Need to fix some other issues

**Comment by fcassia2017** [comment, marked as answer]
Wow great turnaround time. 💪💪👍👍👍
Thank you so much
